datastore.go

```diff
@@ -0,0 +1,201 @@
+// This file is part of the go-meta library.
+//
+// Copyright (C) 2017 JAAK MUSIC LTD
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// If you have any questions please contact yo@jaak.io
+
+package meta
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sync/syncmap"
+
+	"github.com/ethereum/go-ethereum/swarm/api/client"
+	multihash "github.com/multiformats/go-multihash"
+)
+
+// Datastore represents storage for any key-value pair.
+type Datastore interface {
+	put(data []byte) (mhash multihash.Multihash, err error)
+	get(key string) (value []byte, err error)
```

Comment:
We should document that `key` should be the hex encoding of the hash part of the multihash returned from `put` (that isn't immediately obvious)